Kernel/Net/NetworkAdapter.cpp

```diff
@@ -69,6 +69,34 @@ void NetworkAdapter::fill_in_ipv4_header(PacketWithTimestamp& packet, IPv4Addres
     ipv4.set_checksum(ipv4.compute_checksum());
 }
 
+void NetworkAdapter::fill_in_ipv6_header(PacketWithTimestamp& packet, IPv6Address const& source_ipv6, MACAddress const& destination_mac, IPv6Address const& destination_ipv6, TransportProtocol protocol, size_t payload_size, u8 hop_limit)
+{
+    size_t ipv6_packet_size = sizeof(IPv6PacketHeader) + payload_size;
+    VERIFY(ipv6_packet_size <= mtu());
+
+    size_t ethernet_frame_size = ipv6_payload_offset() + payload_size;
+    VERIFY(packet.buffer->size() == ethernet_frame_size);
+    memset(packet.buffer->data(), 0, ipv6_payload_offset());
+    auto& eth = *bit_cast<EthernetFrameHeader*>(packet.buffer->data());
+    eth.set_source(mac_address());
+    eth.set_destination(destination_mac);
+    eth.set_ether_type(EtherType::IPv6);
+    auto& ipv6 = *bit_cast<IPv6PacketHeader*>(eth.payload());
+    ipv6.set_version(6);
+    ipv6.set_destination(destination_ipv6);
+    ipv6.set_source(source_ipv6);
+    switch (protocol) {
+    case TransportProtocol::ICMPv6:
+        ipv6.set_next_header(static_cast<u8>(TransportProtocol::ICMPv6));
+        break;
+    default:
+        dbgln_if(ICMPV6_DEBUG, "fill_in_ipv6_header: Unknown TransportProtocol, setting NoNextHeader");
+        ipv6.set_next_header(static_cast<u8>(IPv6NextHeader::NoNextHeader));
+    }
+    ipv6.set_length(payload_size);
+    ipv6.set_hop_limit(hop_limit);
```

Comment:
Personally I'd like to see this code chunked into a few more logical blocks, for readability sake